### ISSUE TITLE
Questions about the weight

### ISSUE THREAD

**Issue by cbontoiu** [ComputationalRadiationPhysics/picongpu#4777]
Hi,

I understand that PIConGPU was recently updated so ```uy, ux, uz``` don't need to be multiplied by the weight ```w``` anymore.

**1.** This is more a question of Relativity as applied to the ```openPMD``` format. Is there a way to express x- and z- momenta in units of [eV/c] starting from equation (14) in the snapshot shown below? I can get the kinetic energy in units of [eV] but not the transverse momenta in units of [eV/c].

![image](https://github.com/ComputationalRadiationPhysics/picongpu/assets/7108691/73607c43-95de-47bd-91bf-3736637d2aff)

**2.** Working with electron macroparticles, assuming that I want to calculate the total charge of a collection, is it still correct to multiply the _weight_ ```w``` of each macroparticle by the electron charge ```e``` and sum up all products, irrespective of the particle shape _order_?
 
```
charge = 0; 
for j ....:
     charge += w[j] * e;
```
In other words, for example, is the real charge squared of what is calculated above for a ```particles::shapes::TSC```, or the  particle shape _order_ doesn't matter here at all?

**3.** I normally obtain the electron charge density using the ```openPMD Viewer``` as

   ```rho, info = ts.get_field(iteration = it, field = 'e_all_chargeDensity', slice_across='z', slice_relative_position=0, plot=False);```

Does this come in SI units, that is [C/m^3] as I assumed so far, or do I need to use _weight_ and particle shape _order_?

I'm working with:
```
openPMD: 0.15.2
PIConGPU: 0.7.0
openpmd-viewer: 1.9.0
```

Thanks a lot for your help.
Cristian

**Comment by psychocoderHPC** [comment]
> I understand that PIConGPU was recently updated so uy, ux, uz don't need to be multiplied by the weight w anymore.

I think you are pointing to this BUG-FIX https://github.com/ComputationalRadiationPhysics/picongpu/pull/4636
PIConGPU always stored the momentum with the corresponding weighting included but due to a missing attribute for OpenPMD is was not correctly handled in the OpenPMD viewer.

> 1. This is more a question of Relativity as applied to the openPMD format. Is there a way to express x- and z- momenta in units of [eV/c] starting from equation (14) in the snapshot shown below? I can get the kinetic energy in units of [eV] but not the transverse momenta in units of [eV/c].

I do not think we have a predefined derived attribute implementation to do it. Let me ping @steindev @PrometheusPi 

for 2. and 3. I hope that @steindev @PrometheusPi can help you too.

**Comment by cbontoiu** [comment]
Hi @psychocoderHPC, @steindev  and @PrometheusPi 
 
My main worry for question **1** is the following. Is it correct to plugin ```uy``` from (17) into (18)? This is because (17) stands for a macroparticle but (18) refers to electron rest energy so I am putting equal between the macroparticle momentum and the electron momentum, although physically one macroparticle stands for more electrons. I consider the macroparticle as a sampling device (probe) for the average momentum of all nearby electrons. This analogy is clear when calculating the charge through the weight but not when calculating the momentum. Do I need to handle the weight somehow? 

Questions **2** and **3** are well formulated I think. Please let me know if you can help. I appreciate.

Cristian

**Comment by BrianMarre** [comment]
*regarding 1.)*

- equation (16) is wrong in the general case.
```
gamma^2 - 1 = beta^2 /( 1 - beta^2 )
sqrt(gamma^2 - 1) = | \vec(beta) | / sqrt(1 - beta^2)
sqrt(gamma^2 - 1) / gamma = | \vec(beta) | / sqrt(1 - beta^2) * sqrt(1 - beta^2) 
sqrt(gamma^2 - 1) / gamma = | \vec(beta) | =/= beta_y
```
equation (17) depends on component level information on beta, as such you can not substitute the vector magnitude from the corrected equation (16)

to get \vec(u) from the openPMD-output you may use the following formulas:
```
\vec(beta) = \vec(p) / (p^2 + m^2 * c^2)
gamma = sqrt(1 + p^2 / (m^2 * c^2))

\vec(u) = beta * gamma
```

in these you may directly use the openPMD values since both m and p are weighted according to the openPMD standard, [here](https://github.com/openPMD/openPMD-standard/blob/f2b5edc32640997d904cd3366137a3f9f135e288/EXT_ED-PIC.md?plain=1#L397-L403) and [here](https://github.com/openPMD/openPMD-standard/blob/f2b5edc32640997d904cd3366137a3f9f135e288/EXT_ED-PIC.md?plain=1#L413-L420), and there weights exactly cancel.
Kinetic energy should be straight forward from gamma.

*regarding 2.)*

For a collection of electrons you are correct. For ions, you have to take the possibly varying charge state into account.
To get the charge of a collection of electrons from openPMD output, just add the charge of all macro particles up and remember to multiply by unitChargeSI if you need it in SI.

charge is weighted according to the [openPMD-standard](https://github.com/openPMD/openPMD-standard/blob/f2b5edc32640997d904cd3366137a3f9f135e288/EXT_ED-PIC.md?plain=1#L389-L395)

For PIConGPU we by default use (-e)*(typical NumberParticlesPerMacroParticle) as the internal unit for charge, see [here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/5456d097efc3c6e60bb395e11ff452a3b280963e/include/picongpu/param/unit.param#L50), but this is user configurable. so do not rely on that. The openPMD standard guarantees are better.

shape is only a factor if you consider densities/spatial distribution, since they integrating them over the entire space by definition yields 1

regarding 3) sorry I am not using openPMD-viewer myself, so I do not know ;)

**Comment by BrianMarre** [comment]
if you are unsure about reading openPMD data using openPMD-api, take a look [here](https://github.com/BrianMarre/picongpuAtomicPhysicsTools/blob/40ef0357ef8e2d717cd5c0b41b0f627b025f59a2/ReadingOutputFromFLYonPIC_v2/Reader/openPMD.py) for an application example I wrote.

**Comment by cbontoiu** [comment]
@BrianMarre thanks for your detailed explanations!

OK, apart from understanding the equations I want to safely convert from what PIConGPU gives me for a macroparticle to the kinetic energy associated with that macroparticle.

From your equation:

```gamma = sqrt(1 + p^2 / (m^2 * c^2))```

it is tempting to replace ```p^2``` by ```ux^2``` + ```uy^2``` + ```uz^2``` where ```ux```, ```uy```, ```uz``` come from _openPMD_ in units of ```[m_ec]``` This would give me ```gamma``` and therefore the kinetic energy for the macroparticle.

_Example:_
```
# openPMD data

uy = 2.691728e+01 #[m_e*c]
ux = -3.815856e-01 #[m_e*c]
uz = -2.309816e+00 #[m_e*c]

gamma = sqrt(1 + pow(ux, 2) + pow(uy, 2) + pow(uz, 2)) # 27.04

E0 = 0.511; # [MeV] electron rest energy
Ekin = (gamma - 1)*E0 # 13.3 MeV
```
Would this be correct? 

If yes, there is one more thing to clarify. If a macroparticle stands for more real electrons does the macroparticle```gamma``` found above, stand for the average ```gamma``` among all nearby electrons (as controlled by the weight and the shape function, details which are not important for the current calculation)? 

Many thanks.
Cristian

**Comment by BrianMarre** [comment]
yes your kinetic energy is correct.

Regarding the interpretation of gamma:
A macro particle is a phase space sample, therefore the calculated gamma would be the gamma of a physical particle this phase space sample corresponds to, not really an average gamma over a cloud of particles, since by definition all particles corresponding to a single macro particle have the exact same velocity, they only differ in position.

In general it is not a good idea to consider macro particles as a cloud of physical particles, but rather only as samples of phase space density.

**Comment by PrometheusPi** [comment]
Sorry for responding so late @cbontoiu and thanks for taking over this issue @BrianMarre. 

Please be aware that openPMD-viewer returns unitless momentum $u_i = \beta_i * \gamma = p_i / (m_e c w)$ while most equations assume real momentum (containing the mass) thus expressing p as sum of u_i is not correct. However, openPMD-viewer's unitless definition does not depend on the weighting anymore.

@cbontoiu Did @BrianMarre answer all your questions (and can we close this issue)?

**Comment by cbontoiu** [comment]
@PrometheusPi yes, it is clear. Thank you. I will close this issue.
